L1Trigger/L1TMuonEndCapPhase2/src/DAQ/CSCTPConverter.cc

```diff
@@ -0,0 +1,181 @@
+  const int tp_subsector = tp_info.subsector;
+  const int tp_chamber = tp_info.chamber;
+  const int tp_station = tp_info.station;
+  const int tp_ring = tp_info.ring;
+  const int tp_layer = tp_info.layer;
+
+  const int tp_csc_id = tp_info.csc_id;
+  const auto tp_csc_facing = tp_info.csc_facing;
+
+  // Unpack data
+  const int tp_strip = tp_data.strip;
+  const int tp_strip_quart_bit = tp_data.strip_quart_bit;
+  const int tp_strip_eighth_bit = tp_data.strip_eighth_bit;
+  const int tp_strip_quart = tp_data.strip_quart;
+  const int tp_strip_eighth = tp_data.strip_eighth;
+
+  const int tp_wire1 = tp_info.csc_first_wire;
+  const int tp_wire2 = tp_info.csc_second_wire;
+
+  int tp_bend;
+  const int tp_slope = tp_data.slope;
+
+  const int tp_bx = tp_info.bx;
+  const int tp_subbx = 0;   // no fine resolution timing
+  const float tp_time = 0;  // no fine resolution timing. Note: Check with Efe, Jia Fu gets this from digi directly.
+
+  const auto tp_selection = tp_info.selection;
+
+  const int tp_pattern = tp_data.pattern;
+  const int tp_quality = 6;
+
+  // Apply CSC Run 2 pattern -> bend conversion
+  // Override tp_bend
+  constexpr int tp_bend_lut_size = 11;
+  constexpr int tp_bend_lut[tp_bend_lut_size] = {-5, 5, -4, 4, -3, 3, -2, 2, -1, 1, 0};
+  emtf_assert(tp_pattern < tp_bend_lut_size);
+  tp_bend = tp_bend_lut[tp_pattern];
+  tp_bend *= tp_endcap_pm;  // sign flip depending on endcap
+
+  // ID scheme used in FW
+  const int tp_ilink = tp_info.ilink;
+
+  // Get Global Coordinates
+  const GlobalPoint& gp_w1 = GEOM.getGlobalPoint(tp);
+  const float glob_phi_w1 = tp::rad_to_deg(gp_w1.phi().value());
+  const float glob_theta_w1 = tp::rad_to_deg(gp_w1.theta().value());
+  const double glob_rho_w1 = gp_w1.perp();
+  const double glob_z_w1 = gp_w1.z();
+
+  // Calculate EMTF Values
+  const int emtf_phi_w1 = tp::calc_phi_int(sector_, glob_phi_w1);
+  const int emtf_bend_w1 = std::clamp(tp_bend * 4, -16, 15);  // 5-bit, signed
+  const int emtf_theta_w1 = tp::calc_theta_int(tp_endcap_pm, glob_theta_w1);
+  const int emtf_qual_w1 = std::clamp(tp_quality, 0, 15);  // 4-bit, unsigned
+  const int emtf_site_w1 = context_.site_lut_.lookup({tp_subsystem, tp_station, tp_ring});
+  const int emtf_host_w1 = context_.host_lut_.lookup({tp_subsystem, tp_station, tp_ring});
+  const int emtf_zones_w1 = context_.zone_lut_.get_zones(emtf_host_w1, emtf_theta_w1);
+
+  // Calculated Ambiguous Info
+  int emtf_theta_w2 = 0;
+  int emtf_qual_w2 = tp_pattern;
+
+  if (tp_wire2 > -1) {
+    auto tp_w2 = tp;
+
+    tp_w2.accessCSCData().keywire = tp_wire2;
+
+    const GlobalPoint& gp_w2 = GEOM.getGlobalPoint(tp_w2);
+    const double glob_theta_w2 = tp::rad_to_deg(gp_w2.theta().value());
+
+    emtf_theta_w2 = tp::calc_theta_int(tp_endcap_pm, glob_theta_w2);
+  }
+
+  emtf_assert((0 <= emtf_phi_w1) and (emtf_phi_w1 < 5040));
+  emtf_assert((1 <= emtf_theta_w1) and (emtf_theta_w1 < 128));
+  emtf_assert((0 <= emtf_theta_w2) and (emtf_theta_w2 < 128));
```

Comment:
It's ok for emtf_assert to remain in the code. The current Run-3 EMTF emulator also uses a similar approach so that we can have assert during tests. It's disabled in production and is harmless.